Source/WebCore/bindings/js/JSMediaSessionCustom.cpp

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (C) 2023 Apple, Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "JSMediaSession.h"
+
+#include <JavaScriptCore/JSCInlines.h>
+
+namespace WebCore {
+template <typename Visitor>
+void JSMediaSession::visitAdditionalChildren(Visitor& visitor)
+{
+    for (auto& actionHandler : wrapped().actionHandlers()) {
```

Comment:
@rniwa @Constellation Do we need Lock for actionHandlers here? I believe visitAdditionalChildren() gets called off the main thread without synchronization so I suspect it isn't safe to iterate this map without locking but I am not 100% sure.

Reply:
This function can be called concurrently while main thread is working. So if this map can be changed after the initialization, then it should be guarded by the lock.

Reply:
Yeah, we need a lock.

Reply:
The actionHandlers map seems to change dynamically so we need a lock. My proposal would be to add a `visitActionHandlers()` to MessageSession and have it grab a lock while iterating the map. Then call this visit function from here.

Reply:
I copied the pattern I saw for other HashMaps that have locks. Thanks!